src/crimson/os/memstore/MemStore.h

```diff
@@ -0,0 +1,412 @@
+  int _omap_setkeys(const coll_t& cid, const ghobject_t &oid, bufferlist& aset_bl);
+  int _omap_rmkeys(const coll_t& cid, const ghobject_t &oid, bufferlist& keys_bl);
+  int _omap_rmkeyrange(const coll_t& cid, const ghobject_t &oid,
+		       const string& first, const string& last);
+  int _omap_setheader(const coll_t& cid, const ghobject_t &oid, const bufferlist &bl);
+
+  int _collection_hint_expected_num_objs(const coll_t& cid, uint32_t pg_num,
+      uint64_t num_objs) const { return 0; }
+  int _create_collection(const coll_t& c, int bits);
+  int _destroy_collection(const coll_t& c);
+  int _collection_add(const coll_t& cid, const coll_t& ocid, const ghobject_t& oid);
+  int _collection_move_rename(const coll_t& oldcid, const ghobject_t& oldoid,
+			      coll_t cid, const ghobject_t& o);
+  int _split_collection(const coll_t& cid, uint32_t bits, uint32_t rem, coll_t dest);
+  int _merge_collection(const coll_t& cid, uint32_t bits, coll_t dest);
+
+  int _save();
+  int _load();
+
+  void dump(Formatter *f);
+  void dump_all();
+
+public:
+  MemStore(StoreContext *scct, const string& path)
+    : ObjectStore(NULL, path),
+      scct(scct),
+      coll_lock("MemStore::coll_lock"),
+      finisher(scct),
+      used_bytes(0) {}
+  ~MemStore() override {} 
+
+  string get_type() override {
+    return "memstore";
+  }
+
+  bool test_mount_in_use() override {
+    return false;
+  }
+
+  int mount() override;
+  int umount() override;
+
+  int fsck(bool deep) override {
+    return 0;
+  }
+
+  int validate_hobject_key(const hobject_t &obj) const override {
+    return 0;
+  }
+  unsigned get_max_attr_name_length() override {
+    return 256;  // arbitrary; there is no real limit internally
+  }
+
+  int mkfs() override;
+  int mkjournal() override {
+    return 0;
+  }
+  bool wants_journal() override {
+    return false;
+  }
+  bool allows_journal() override {
+    return false;
+  }
+  bool needs_journal() override {
+    return false;
+  }
+
+  int get_devices(set<string> *ls) override {
+    // no devices for us!
+    return 0;
+  }
+
+  int statfs(struct store_statfs_t *buf) override;
+
+  bool exists(CollectionHandle &c, const ghobject_t& oid) override;
+  int stat(CollectionHandle &c, const ghobject_t& oid,
+	   struct stat *st, bool allow_eio = false) override;
+  int set_collection_opts(
+    CollectionHandle& c,
+    const pool_opts_t& opts) override;
+  int read(
+    CollectionHandle &c,
+    const ghobject_t& oid,
+    uint64_t offset,
+    size_t len,
+    bufferlist& bl,
+    uint32_t op_flags = 0) override;
+  using ObjectStore::fiemap;
+  int fiemap(CollectionHandle& c, const ghobject_t& oid,
+	     uint64_t offset, size_t len, bufferlist& bl) override;
+  int fiemap(CollectionHandle& c, const ghobject_t& oid, uint64_t offset,
+	     size_t len, map<uint64_t, uint64_t>& destmap) override;
+  int getattr(CollectionHandle &c, const ghobject_t& oid, const char *name,
```

Comment:
sure, I will add the encapsulation next step.

Reply:
\<thinking loudly mode on />

We might want to consider keeping the `ObjectStore` untouched and put the futurized variants into new interface `FuturizedObjectStore` (need to reconsider the naming :-)):

```cpp
#ifdef WITH_SEASTAR
// intentionally not inheriting from ObjectStore to keep this small & clean
class FuturizedObjectStore {
public:
  // let's stick to bl at least on initial stages of development
  using buffercont = ceph::bufferlist;

  virtual seastar::future<buffercont> read(uint64_t offset, uint64_t len) = 0;
  // ...
};
#endif
```
Then `FuturizedMemStore` would be:
```cpp
#ifdef WITH_SEASTAR
class FuturizedMemStore : private MemStore, public FuturizedObjectStore {
public:
  using MemStore::really_necessary_things;

  seastar::future<buffercont> read(uint64_t offset, uint64_t len) override;
};
#endif
```

I personally imagine the alien `ObjectStore` encapsulator as an adapter:
```cpp
#ifdef WITH_SEASTAR
class AdapterNeedingBetterName : public FuturizedObjectStore;
  ObjectStore& os;
  ceph::thread::ThreadPool tp;
  // ...

  seastar::future<buffercont> read(uint64_t offset, uint64_t len) {
    // execute `int read(off, len, bl)` on `os` via `tp`.
    // Repack `bl`/inject deleter? This could be the painful part
  }
};
#endif
```

What do you think?

Reply:
Hmm, **just in the case** we would be enforced to support some hardware wanting to eat more than 1 core/have a RAID/have few connections/have few CRUSH leafs (messenger is be reused, so we'd be doing aggregation from RADOS' point of view):

```cpp
#ifdef WITH_SEASTAR
// FuturizedObjectStore implementation for sharding requests across other
// FuturizedObjectStore instances. It would encapsulate the *crossbar* while
// being jettisonable. It brings flexibility without the too much costs for
// main deployment scenario (1 OSD/1 core only).
// It can be implemented when it becomes necessary. Definitely after
// FuturizedMemStore.
class ShardingFuturizedObjectStore : public FuturizedObjectStore;
  seastar::sharded<FuturizedObjectStore> sharded_os_collection;
  // ...

  seastar::future<buffercont> read(
     CollectionHandle &c,
     const ghobject_t& oid,
     uint64_t offset,
     size_t len,
     uint32_t op_flags = 0) {
    auto shard_id = pickup_shard(c, oid); // or just c
    return sharded_os_collection.invoke_on(shard_id, [] (auto os_shard) {
       return os_shard->read(...).finally(do_repacking);
      });
  }
};
#endif
```

This is very, very early idea. Need to think more on it.

CC: @tchaikov.

Reply:
if we plan to support sharded cirmson-osd (PG also sharded) and each shard has its own objectstore instance, then we need consider cross shard objectstore instance access in case pg changes. if all crimison-osd shards share the same one ObjectStore instance, then no need this kind cross shard access,  It depends on our finally design, so just keep it in mind.

Reply:
@tchaikov @rzarzynski another option is we can use seastar::thread (seastar::async) instead of posix thread to execute some block operation such as io, that seastar::thread can tolerate blocking. that seastar::thread is different from posix thread, it is not preempted until it is blocked. and running on the same core when launch it. so needn't to do alien and seastar communication.

Reply:
> if we plan to support sharded cirmson-osd (PG also sharded) and each shard has its own objectstore instance, then we need consider cross shard objectstore instance access in case pg changes. if all crimison-osd shards share the same one ObjectStore instance, then no need this kind cross shard access, It depends on our finally design, so just keep it in mind.

@liu-chunmei: the question (already asked on ceph-devel) is what we really understand under the *OSD* concept. I've got an impression it still not answered entirely. Let me ask:

Is `ObjectStore` **interface** (same would apply to `FuturizedObjectStore`) is a part of *OSD*?
Is **an implementation**  of `ObjectStore` a part of `OSD`?

Personally I think the **interface** fits inside *OSD* but its **implementation** not. Differences in answers can be source of many red herrings with the *lack of flexibility* being one of them. I could imagine very simple, single core *OSD* (crimson-osd) without:
* `seastar::sharded`,
* `seastar::foreign_ptr`,
* the entire *crossbar* concept

consuming a complex **implementation** of `FuturizedObjectStore` that makes use of all those things! This would basically be its *implementation detail*. Moreover, such implementation could be:

* developed separately,
* interchangeable (especially with a very simple one like the `FuturizedMemStore` ).

I need to take some sleep, rethink and refine the terminology. :-)

Reply:
@rzarzynski agree, you want a perfect & complex implementation which will adapt for any use case, my thinking is we can have a simple implementation just adapt for single core crimson-osd at the beginning, they are not contradict, just verify our single core crimson-osd first then  if we get positive result we can extend the code to support multiple use cases, so we keep the idea in mind first and implement it later.

Reply:
> EDIT: this interchangeability of implementation (and thus the abstraction of crossbar) can be the way to unify the simplicity and flexibility camps.

yes. i am all for it.

> another option is we can use seastar::thread (seastar::async) instead of posix thread to execute some block operation such as io, that seastar::thread can tolerate blocking.

by "blocking", i think what you mean is the async calls wrapped by seastar, instead of the blocking calls offered by glibc, right?

Reply:
@tchaikov you are right, not the blocking system call, should be blocked at future.get(), so we can't use it directly for objectstore block calls, I misunderstand it before.